sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java

```diff
@@ -164,12 +166,40 @@ public boolean matches(String fileName) {
 
       public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
         throws IOException {
-        ZipInputStream zip = new ZipInputStream(Channels.newInputStream(channel));
+        FullZipInputStream zip = new FullZipInputStream(Channels.newInputStream(channel));
         return Channels.newChannel(zip);
       }
     };
 
     /**
+     * Extend of {@link ZipInputStream} to automatically read all entries in the zip.
+     */
+    public class FullZipInputStream extends ZipInputStream {
+
+      private ZipEntry currentEntry;
+
+      public FullZipInputStream(InputStream is) throws IOException {
+        super(is);
+        currentEntry = getNextEntry();
+      }
+
+      public int read(byte[] b, int off, int len) throws IOException {
```

Comment:
there are a lot of variants of `read` in an `InputStream`. Is it obvious that this is the only variant you need to override? (May well be, I just don't know).